server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.pattern.ConverterKeys;
+import org.apache.logging.log4j.core.pattern.LogEventPatternConverter;
+import org.apache.logging.log4j.core.pattern.PatternConverter;
+import org.apache.lucene.util.CloseableThreadLocal;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.util.LazyInitializable;
+
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+@Plugin(category = PatternConverter.CATEGORY, name = "NodeAndClusterIdConverter")
+@ConverterKeys({"node_and_cluster_id"})
+public final class NodeAndClusterIdConverter extends LogEventPatternConverter implements ClusterStateListener {
```

Comment:
I think rather than making the pattern converted a cluster state listener, I might make a cluster state listener that writes to a static method in the pattern converted. That'd keep Elasticsearch concepts out of the log4j extension code and wouldn't force leaking "funny" log4j2 constructs like `newInstance` into Elasticsearch.

Reply:
Ineed that would be a good design choice to separate the cluster state listener and the pattern converter. 
As discussed the static atomicReference field is safe here, as we only expect one instance of PatternConverter per jvm.